revert "fix wrong output when using jsonpath"

atiratree opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug
/kind regression

#### What this PR does / why we need it:
This partially reverts commit #98057

the original problem was caused by not using {end} at the end of the range. Please see attached jsonpath documentation on using range.

The reverted commit causes a new bug of not parsing missing keys incorrectly (added a unit test for this use case)

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #104276

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fix a 1.22 regression with wrong output using jsonpath
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[Other doc]: https://kubernetes.io/docs/reference/kubectl/jsonpath/
```

atiratree committed:
revert "fix wrong output when using jsonpath"

This partially reverts commit 39cfe232325d66bcdbc935af7aaf7022562e7010and PR #98057

the original problem was caused by not using {end} at the end of the range

atiratree commented:
@zerodayz @soltysh

caesarxuchao commented:
/assign @caesarxuchao @soltysh

knight42 reviewed (approved):
/ok-to-test
/lgtm

Thanks for the fix!

knight42 commented:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/104172/pull-kubernetes-integration/1423531656039370752

These errors doesn't seem to be legit.
/retest

soltysh reviewed (approved):
/lgtm
/approve
@atiratree can you open a cherry-pick against 1.22 to fix this problem since that's where the regression was introduced.

soltysh commented:
/triage accepted
/priority important-longterm

soltysh commented:
ping @caesarxuchao for approval

sttts commented on staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go:
```diff
@@ -806,8 +807,8 @@ func TestRunningPodsJSONPathOutput(t *testing.T) {
-				"when range is used in a certain way in script, additional line is printed",
-				`{range .items[?(.status.phase=="Running")]}{.metadata.name}{" is Running\n"}`,
+				"range over pods without selecting the last one",
+				`{range .items[?(.status.phase=="Running")]}{.metadata.name}{" is Running\n"}{end}`,
```
shouldn't we have both cases tested, ie with and without end?

atiratree replied:
Should we test invalid syntaxes here? What should be the behaviour?
- the output is broken and outputs additional line (current behaviour)
- or should we fast fail with invalid syntax error? (would require changes)

sttts replied:
it was kind of graceful before to leave out the `{end}`, just a little wrong. Could be get it working easily too?

sttts replied:
In any case, I would rather try to avoid breaking behaviour that ran without error before. Fixing it is ok, but not suddenly fail.

atiratree replied:
I am not sure how easy that would be. 
But do we want to support an invalid syntax? I think it is a bit problematic of supporting this especially when you consider nested ranges. I would prefer an error here , but that would also break some scripts (more than keeping it failing slightly).

soltysh replied:
I think that the support for missing `{end}` was rather accidental than intentional. I'd prefer we stick with having code simple and complying to jsonpath standard, than over-complicated to support something not entirely correct.

sttts replied:
I would prefer keeping graceful behaviour and not fix it and break users. But I am not blocking this. Both paths are acceptable.

sttts commented on staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go:
```diff
@@ -280,6 +280,7 @@ func TestStructInput(t *testing.T) {
 
 	missingKeyTests := []jsonpathTest{
 		{"nonexistent field", "{.hello}", storeData, "", false},
+		{"nonexistent field 2", "before-{.hello}after", storeData, "before-after", false},
```
how is this related? I don't see it in the reverted PR.

atiratree replied:
this is a behaviour that got broken by the reverted commit. Ie the template got discarded after the `{.hello}` element.

atiratree committed:
add a test for jsonpath template parsing to prevent regressions

This behaviour was broken by commit
39cfe232325d66bcdbc935af7aaf7022562e7010 and PR kubernetes#98057

soltysh reviewed (approved):
/lgtm

sttts commented:
/approve

with https://github.com/kubernetes/kubernetes/pull/104172#discussion_r699239375 in mind.

atiratree commented:
the failures don't seem connected to this PR
/retest

rtheis commented:
If not too much trouble, can we link this PR as the fix for https://github.com/kubernetes/kubernetes/issues/104276? Thank you.

soltysh commented:
> If not too much trouble, can we link this PR as the fix for #104276? Thank you.

done
